File: src/haskell/Stack.hs
module Stack where

data Stack a = Empty | Cons a (Stack a)

-- Inicializa uma pilha com o valor para pilhas vazias.
init :: Stack a
init = Empty

-- Obtém a cabeça da pilha, o primeiro elemento.
head :: Stack a -> Maybe a
head Empty      = Nothing
head (Cons x _) = Just x

-- Obtém a cauda da pilha, todos os elementos que não o primeiro.
tail :: Stack a -> Maybe (Stack a)
tail Empty       = Nothing
tail (Cons _ xs) = Just xs

-- Checa se a pilha está vazia.
isEmpty :: Stack a -> Bool
isEmpty Empty = True
isEmpty _     = False

-- Remove o elemento da Stack.
-- Retorna numa tupla o primeiro elemento e uma pilha com os outros elementos.
pop :: Stack a -> (Maybe a, Maybe (Stack a))
pop stack = (Stack.head stack, Stack.tail stack)

-- Adiciona um elemento na Stack.
push :: Stack a -> a -> Stack a
push stack value = Cons value stack

-- Obtém o primeiro elemento da Stack.
peek :: Stack a -> Maybe a
peek = Stack.head

File: src/typescript/Stack.ts
type Stack<T> = null | Readonly<{
  head: T;
  tail: Stack<T>;
}>;

// Inicializa uma pilha com o valor para pilhas vazias.
const init = <T>(): Stack<T> => null;

// Obtém a cabeça da pilha, o primeiro elemento.
const head = <T>(stack: Stack<T>): null | T =>
  stack === null ? null : stack.head;

// Obtém a cauda da pilha, todos os elementos que não o primeiro.
const tail = <T>(stack: Stack<T>): Stack<T> =>
  stack === null ? null : stack.tail;

// Checa se a pilha está vazia.
const isEmpty = <T>(stack: Stack<T>): boolean => stack === null;

// Remove o elemento da Stack.
// Retorna numa tupla o primeiro elemento e uma pilha com os outros elementos.
const pop = <T>(stack: Stack<T>): [null | T, Stack<T>] => [
  head(stack),
  tail(stack)
];

// Adiciona um elemento na Stack.
const push = <T>(stack: Stack<T>, value: T): Stack<T> =>
  Object.create(null, {
    head: {
      value,
      writable: false,
      enumerable: true,
      configurable: false
    },
    tail: {
      value: stack,
      writable: false,
      enumerable: true,
      configurable: false
    }
  });

// Obtém o primeiro elemento da Stack.
const peek = head;

export { Stack, init, head, tail, isEmpty, pop, push, peek };
